dangerzone/gui/updater.py

```diff
@@ -0,0 +1,225 @@
+"""A module that contains the logic for checking for updates."""
+
+import logging
+import platform
+import typing
+from typing import Any, Dict, Optional
+
+if typing.TYPE_CHECKING:
+    from PySide2 import QtCore
+else:
+    try:
+        from PySide6 import QtCore
+    except ImportError:
+        from PySide2 import QtCore
+
+import markdown
+import requests
+
+from ..util import get_version
+from .logic import Alert, DangerzoneGui
+
+log = logging.getLogger(__name__)
+
+
+# TODO: Create a wiki page that explains how updates work for Dangerzone, before
+# merging this.
+MSG_CONFIRM_UPDATE_CHECKS = """\
+Do you want to be notified about new Dangerzone releases?
+
+If "Yes", Dangerzone will check GitHub for new releases on startup. If "No", Dangerzone will make no network requests and won't inform you about new releases.
+
+If you prefer another way of getting notified about new releases, we suggest adding to your RSS reader our Mastodon feed: https://fosstodon.org/@dangerzone.rss. For more information about updates, check our wiki page: https://github.com/freedomofpress/dangerzone/wiki/Updates
+"""
+
+
+class Updater(QtCore.QThread):
+    """Check asynchronously for Dangerzone updates.
+
+    The Updater class is mainly responsible for the following:
+
+    1. Asking the user if they want to enable update checks or not.
+    2. Determining when it's the right time to check for updates.
+    3. Hitting the GitHub releases API and learning about updates.
+
+    Since checking for updates is a task that may take some time, we perform it
+    asynchronously, in a Qt thread. This thread then triggers a signal, and informs
+    whoever has connected to it.
+    """
+
+    finished = QtCore.Signal(dict)
+
+    GH_RELEASE_URL = (
+        "https://api.github.com/repos/freedomofpress/dangerzone/releases/latest"
+    )
+
+    def __init__(self, dangerzone: DangerzoneGui):
+        super().__init__()
+        self.dangerzone = dangerzone
+
+    ###########
+    # Helpers for updater settings
+    #
+    # These helpers make it easy to retrieve specific updater-related settings, as well
+    # as save the settings file, only when necessary.
+
+    def _save_me_maybe(self, key: str, val: Any) -> None:
```

Comment:
Yeap, I considered it. I checked the rest of the code if it would benefit from this change, and what I found out is that we **currently** tend to set several keys, and then save them once and explicitly, in a transactional fashion. For example:

https://github.com/freedomofpress/dangerzone/blob/e989069712e41f4721aa2f1fba4e49d70cded7ef/dangerzone/gui/main_window.py#L642-L657

So, what we can do is add a `save=False` keyword argument in `Settings.set()`, and let the updater code use it, whereas the existing code will not use it. Are you ok with this?

Reply:
I had not noticed the transaction-style save settings. Looking at the source for `set(` it looks like all `set()` are followed by a `save()` Do you think it's critical to be this way? I think it may create more bugs that what it solves. For example:  we may forget to save it somewhere.

What to do you think about ditching the `save()` and move your `_save_maybe` to the `set()` method? That way we save on some writes, but don't run the risk of forgetting to save.

Reply:
I wouldn't worry that much about it to be honest. This transactional logic exists in two places tops, and it's been there for a while. If anything, I'd prefer tackling it in a later PR.

Reply:
OK. So we leave it like this?

Reply:
I can definitely add the `save=False` logic, but we can change the logic in `main_window.py` in a subsequent PR.

Reply:
Should be fixed in 94a6a13.